sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala

```diff
@@ -28,43 +28,12 @@ object Canonicalize {
       e: Expression,
       f: PartialFunction[Expression, Seq[Expression]]): Seq[Expression] = e match {
     case c if f.isDefinedAt(c) => f(c).flatMap(gatherCommutative(_, f))
-    case other => reorderCommutativeOperators(other) :: Nil
+    case other => other.canonicalized :: Nil
   }
 
   /** Orders a set of commutative operations by their hash code. */
-  private def orderCommutative(
+  def orderCommutative(
       e: Expression,
       f: PartialFunction[Expression, Seq[Expression]]): Seq[Expression] =
     gatherCommutative(e, f).sortBy(_.hashCode())
-
-  def reorderCommutativeOperators(e: Expression): Expression = e match {
-    // TODO: do not reorder consecutive `Add`s or `Multiply`s with different `failOnError` flags
```

Comment:
we need to keep this TODO somewhere.

Reply:
Sure, added back in https://github.com/apache/spark/pull/37851/commits/dead8b66872699023a15944bffd89a16f146bc76